src/future/blocking.rs

```diff
@@ -0,0 +1,95 @@
+use Async;
+use future::Future;
+use executor;
+use task_impl::ThreadNotify;
+
+/// Provides thread-blocking operations on a future.
+///
+/// See [`blocking`] documentation for more details.
+#[derive(Debug)]
+pub struct Blocking<T> {
+    inner: executor::Spawn<T>,
+}
+
+/// Provides thread-blocking operations on a future.
+///
+/// `blocking` consumes ownership of `future`, returning a `Blocking` backed by
+/// the future. The `Blocking` value exposes thread-blocking operations that
+/// allow getting the realized value of the future. For example,
+/// `Blocking::wait` will block the current thread until the inner future has
+/// completed and return the completed value.
+///
+/// **These operations will block the current thread**. This means that they
+/// should not be called while in the context of a task executor as this will
+/// block the task executor's progress.
+pub fn blocking<T: Future>(future: T) -> Blocking<T> {
+    let inner = executor::spawn(future);
+    Blocking { inner }
+}
+
+impl<T: Future> Blocking<T> {
+    /// Query the inner future to see if its value has become available.
+    ///
+    /// Unlike `Future::poll`, this function does **not** register interest if
+    /// the inner future is not in a ready state.
+    ///
+    /// This function will return immediately if the inner future is not ready.
+    pub fn poll(&mut self) -> Option<Result<T::Item, T::Error>> {
```

Comment:
The name `poll` returning something that isn't `Poll` seems confusing, when every other future or stream `poll` will. Maybe something like `peek` or `tap`?

Reply:
I'm open to change the name if a good alternative is proposed.

* `peek` implies the value remains in the future (and there is a peek combinator that does this on Future).
* `tap` I don't understand how tap applies to the behavior (I am not familiar w/ tap as a name here).

Re confusion, the return type is different, so at least it won't compile if you are expecting `Poll`.

Reply:
I want to second @seanmonstar's point about `poll`; we've used that pretty universally to designate that interest *will* be registered.

It could match cases with blocking data structures in `std` and be called `try_wait`, but that's not awesome.

Also, it's a *tad* confusing for this to be part of the `Blocking` wrapper, since the point here is precisely that it *doesn't* block :-)

Reply:
It could be named something besides `Blocking`. The real goal of the type is "Use this future from the edge outside of the future task system", but I went with `Blocking` to make that behavior very explicit.

I considered `try_wait`, but that doesn't really make sense as it isn't trying to wait...

Reply:
The only other name that I thought of would be `try_take`.

Reply:
Oh I'm not complaining about the name `Blocking` (which I like), I'm saying that `Blocking::poll` is a bit confusing because it's *not* in fact blocking. If you're wanting to poll the future outside of an executor, why would you think to reach for `Blocking` to do so?

But anyway, I'm more concerned that we find a name other than `poll`.

Reply:
re: `try_take`, I was thinking along similar lines. Seems good?

Reply:
Does anyone hate `try_take` instead of `poll`?

Reply:
I'm confused as to how this avoids registering interest. My understanding of futures is that they register with the current task before returning `NotReady`. With `poll_future_notify` below, isn't the interest is still registered?